cpp/src/parquet/statistics.cc

```diff
@@ -32,84 +32,362 @@ using arrow::MemoryPool;
+  }
+};
+
+template <>
+struct CompareHelper<ByteArrayType, false> {
+  static inline bool Compare(int type_length, const ByteArray& a, const ByteArray& b) {
+    const uint8_t* aptr = reinterpret_cast<const uint8_t*>(a.ptr);
+    const uint8_t* bptr = reinterpret_cast<const uint8_t*>(b.ptr);
+    return std::lexicographical_compare(aptr, aptr + a.len, bptr, bptr + b.len);
+  }
+};
+
+template <>
+struct CompareHelper<FLBAType, false> {
+  static inline bool Compare(int type_length, const FLBA& a, const FLBA& b) {
+    const uint8_t* aptr = reinterpret_cast<const uint8_t*>(a.ptr);
+    const uint8_t* bptr = reinterpret_cast<const uint8_t*>(b.ptr);
+    return std::lexicographical_compare(aptr, aptr + type_length, bptr,
+                                        bptr + type_length);
+  }
+};
+
+template <typename DType, bool is_signed = true>
+class TypedComparatorImpl : public TypedComparator<DType> {
+ public:
+  typedef typename DType::c_type T;
+
+  explicit TypedComparatorImpl(int type_length = -1) : type_length_(type_length) {}
+
+  bool CompareInline(const T& a, const T& b) {
+    return CompareHelper<DType, is_signed>::Compare(type_length_, a, b);
+  }
+
+  bool Compare(const T& a, const T& b) override { return CompareInline(a, b); }
+
+  void GetMinMax(const T* values, int64_t length, T* out_min, T* out_max) override {
+    T min = values[0];
+    T max = values[0];
+    for (int64_t i = 1; i < length; i++) {
+      if (CompareInline(values[i], min)) {
+        min = values[i];
+      } else if (CompareInline(max, values[i])) {
+        max = values[i];
+      }
+    }
+    *out_min = min;
+    *out_max = max;
+  }
+
+  void GetMinMaxSpaced(const T* values, int64_t length, const uint8_t* valid_bits,
+                       int64_t valid_bits_offset, T* out_min, T* out_max) override {
+    ::arrow::internal::BitmapReader valid_bits_reader(valid_bits, valid_bits_offset,
+                                                      length);
+    T min = values[0];
+    T max = values[0];
+    for (int64_t i = 0; i < length; i++) {
+      if (valid_bits_reader.IsSet()) {
+        if (CompareInline(values[i], min)) {
+          min = values[i];
+        } else if (CompareInline(max, values[i])) {
+          max = values[i];
+        }
+      }
+      valid_bits_reader.Next();
+    }
+    *out_min = min;
+    *out_max = max;
+  }
+
+ private:
+  int type_length_;
+};
+
+std::shared_ptr<Comparator> Comparator::Make(Type::type physical_type,
+                                             SortOrder::type sort_order,
+                                             int type_length) {
+  if (SortOrder::SIGNED == sort_order) {
+    switch (physical_type) {
+      case Type::BOOLEAN:
+        return std::make_shared<TypedComparatorImpl<BooleanType>>();
+      case Type::INT32:
+        return std::make_shared<TypedComparatorImpl<Int32Type>>();
+      case Type::INT64:
+        return std::make_shared<TypedComparatorImpl<Int64Type>>();
+      case Type::INT96:
```

Comment:
Remove?

Reply:
I'm leaving this so it's possible to instantiate the Comparator